### PR TITLE
added timeout to input()

### PR DESCRIPTION
**Pull request by ATMakersBill**
I have added a second optional parameter to input() that acts like a timeout.

When input("PROMPT>", 10) is called with a timeout parameter, it will return in that number of milliseconds if there is no data available on the USB serial connection.  

This is **not** a true async serial read, it's just a workaround to let allow other tasks to occur while waiting for serial input to **start**.  Once a keystroke is read, the method drops into readline processing allowing for command history etc.  And whatever activity is in the loop will stop.

However, this bridges a gap that caused problems for one of our projects. We want to be able to take serial commands over USB to change colors as well as have modes like blink, "throb", "wheel", "cycle" etc. that run between changes.  

With this change, we can do that with the following code:

```
while True:
    inColor = input("", 0)
#result will be None if timeout occurs
    if (inColor is None):
        runMode()
        continue
    if (inColor.lower().startswith("red")):
        targetColor = (255, 0, 0)
        if (mode == "wheel"):
            mode="solid"
    elif (inColor.lower().startswith("green")):
        targetColor = (0, 255, 0)
        if (mode == "wheel"):
            mode="solid"
    elif (inColor.lower().startswith("yellow")):
        targetColor = (200, 200, 0)
        if (mode == "wheel"):
            mode="solid"
    elif (inColor.lower().startswith("@")):
        mode= inColor[1:]
    elif (inColor.startswith("#")):
        hexcode = inColor[1:]
        targetColor = hex2rgb(hexcode)
        if (mode == "wheel"):
            mode="solid"
    else:
        targetColor =(50, 50, 50)
        if (mode == "wheel"):
            mode="solid"
    pixels.fill(targetColor)
    pixels.show()
```
In our case, we used a timeout of zero, but any positive time works.  Sending False is identical to sending 0.  If the parameter is omitted, input() behaves exactly like the original.

An alternative approach (arguably better) would be to add a builtin function that is a passthrough to "usb_bytes_available()".  I'm fine with that approach as well.

I'd love any feedback.

### CODE REVIEW

**Comment by ATMakersBill** [comment]
I see from the Travis build that this fails on some ports that don't have usb_bytes_available().

If there is a desire to go forward, I can add checks for particular architectures or features that work.  For our (short-term) needs the M0 will work for us (which doesn't imply anything for accepting the PR of course - we can build from our fork)

**Comment by tannewt** [comment]
I don't want to add this as is because it breaks compatibility with CPython because input only takes in a prompt. https://docs.python.org/3/library/functions.html#input

You could add this as a custom function in a custom module. It gets it closer to the second usb serial api that would be good in the long term. Let me know if you want help with that.

**Comment by ATMakersBill** [comment]
I get it - I have no problem with that

**Comment by ATMakersBill** [comment]
@tannewt (or whoever), I think I'd like to go ahead and add a module for this.  It's trivial - just one method that checks whether there is input available on the console.  In theory, it should be temporary as the updated async IO code would replace it.  However, I think there are some use cases (including two of mine now) where this one feature is blocking other solutions.

So, the question is where this code should live.  My first thought was under ports/atmel-samd... but then I saw that the ESP8266 (and presumably the ESP32 and NRF) DO have the equivalent of "usb_bytes_available()" just implemented differently.  So this COULD be done cross-platform (with IFDEFs).  I think that means it should go under shared_bindings.

Thoughts?

**Comment by dhalbert** [comment]
Yes, please put something under shared-bindings and if it's not available, then it would throw an error that it's not implemented. Also would be done for the nrf port as well. You could use pyserial as an example of what the API might be. We added stuff to UART that mimics pyserial as well.

File: shared-bindings/supervisor/Runtime.c [hunk 1]
```diff
@@ -80,8 +80,28 @@ const mp_obj_property_t supervisor_serial_connected_obj = {
               (mp_obj_t)&mp_const_none_obj},
 };
 
+/*Added to allow for polling of USB Console*/
```
Comment (tannewt) [inline shared-bindings/supervisor/Runtime.c]: Please add a comment like the one above to document this new attribute. It starts with `//|     .. attribute:: runtime.serial_connected`

**Comment by ATMakersBill** [comment]
I have added the documentation as requested and pushed it to my fork - hopefully that's all that's needed (still figuring out this process)

**Comment by ATMakersBill** [comment]
If there's anything else you need from me let me know asap - I'm headed to NYC for my 20th anniversary :-)

**Comment by tannewt** [comment]
I made one tweak to the location and will merge after travis gives the ok.

**Pull request merged**